### PR TITLE
feat: update cluster secret from metrics

### PR DESCRIPTION
**Pull request by mayzhang2000**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

### CODE REVIEW

File: pkg/apis/application/v1alpha1/types.go [hunk 1]
```diff
@@ -1029,6 +1029,8 @@ type Cluster struct {
 	ServerVersion string `json:"serverVersion,omitempty" protobuf:"bytes,5,opt,name=serverVersion"`
 	// Holds list of namespaces which are accessible in that cluster. Cluster level resources would be ignored if namespace list if not empty.
 	Namespaces []string `json:"namespaces,omitempty" protobuf:"bytes,6,opt,name=namespaces"`
+	// Last Cache sync time
+	LastCacheSyncTime string `json:"lastSyncTime,omitempty" protobuf:"bytes,7,opt,name=lastSyncTime"`
```
Comment (alexmt) [inline pkg/apis/application/v1alpha1/types.go:1033]: `ConnectionState` data structure already has `ModifiedAt` field. Let's use it instead.

File: server/cluster/cluster.go [hunk 2]
```diff
@@ -41,38 +40,30 @@ func NewServer(db db.ArgoDB, enf *rbac.Enforcer, cache *servercache.Cache, kubec
...
 
-	config := cluster.RESTConfig()
-	config.Timeout = time.Second
-	version, err := s.kubectl.GetServerVersion(config)
-	if err != nil {
-		clusterInfo.Status = appv1.ConnectionStatusFailed
-		clusterInfo.Message = fmt.Sprintf("Unable to connect to cluster: %v", err)
-	} else {
-		clusterInfo.Version = version
+	if cluster.LastCacheSyncTime != "" {
```
Comment (alexmt) [inline server/cluster/cluster.go:52]: If the controller is down the cluster connection state will be obsolete. I think we should introduce `Unknown` status in addition to `Successful` and `Failed`. The `Unknown` status should be used is status attribute is missing or `modifiedAt` is too old.

File: controller/metrics/clustersecretupdater.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+			break
+		case <-tick:
+
+			for _, info := range c.infoSource.GetClustersInfo() {
+				cluster, err := c.db.GetCluster(context.Background(), info.Server)
+				if err != nil {
+					continue
+				}
+				cluster.ServerVersion = info.K8SVersion
+				cluster.LastCacheSyncTime = info.LastCacheSyncTime.String()
```
Comment (alexmt) [inline controller/metrics/clustersecretupdater.go:31]: The controller might be unable to connect to cluster. In this case the status should be failed with the appropriate error message. The connection error is not available in ClusterInfo data structure but it is easy to add it:

https://github.com/argoproj/argo-cd/blob/8ad928330f26b3eea523626d306306e4929aaa7a/controller/cache/cluster.go#L675

Lets store both ConnectionState and Message in cluster secret annotations

**Review by alexmt — CHANGES_REQUESTED** [review]

**Comment by mayzhang2000** [comment]
This is replaced by https://github.com/argoproj/argo-cd/pull/3604.

**Pull request closed without merging**